### PR TITLE
Visitor Can Register For An Account

### PR DESCRIPTION
**Pull request by FlisAnn**
[ Visitors can Register for an account  to buy a subscription-Client](https://www.pivotaltracker.com/story/show/176055656)
### User Story
```
As a Visitor,
In order to access unique features on the news app,
I would like to be able to register for an account on the site.
```
## Changes proposed in this pull request:
* Adds a nested modal to include registration and subscription forms
* Adds functionality to registration portion of form
* Refactors `rootReducer` and API request functionality
* Updates UI styling for header menu
## What I have learned working on this feature:
* To make use of semantic UIs nested modal
* Linking `Form` submission functionality from separate component using `props`
* Combining functionality into a single `onClick` / `onSubmit` event
## Packages/Gems added
*`j-tockauth`
## Screenshots (Client)
Screenshot 2021-01-14 at 14 29 43
Screenshot 2021-01-14 at 14 29 30
Screenshot 2021-01-14 at 14 29 13

### CODE REVIEW

File: src/modules/auth.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        type: "SET_CURRENT_USER",
+        payload: response.data.data,
+      });
+    } else {
+      dispatch({
+        type: "SET_ERROR_MESSAGE",
+        payload: "Registration unsuccessful",
+      });
+      localStorage.removeItem("J-tockAuth-Storage");
+    }
```
Comment (emtalen) [inline src/modules/auth.js]: there is no need for this if statement, your default role in the API is registered_user so the role will never be anything different when you sign up

**Review by emtalen — COMMENTED** [review]
I can't see where you are using the `auth` state and therefor it is difficult to see what is not working. We can have a look after the staff meeting

File: src/components/MasterModal.jsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  const [secondOpen, setSecondOpen] = useState(false);
+
+  const registerAndProceed = (e) => {
+    e.preventDefault();
+    performAuthentication(e, dispatch);
+    setSecondOpen(authenticated);
+  };
+
+  const finalizePayment = () => {
+    // Stripe payment functionality will be added here
```
Comment (FlisAnn) [inline src/components/MasterModal.jsx]: Does this comment need to be here?

File: src/components/MasterModal.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      </Button>
+      <Modal
+        onClose={() => setFirstOpen(false)}
+        onOpen={() => setFirstOpen(true)}
+        open={firstOpen}
+        data-cy="first-registration"
+      >
+        <Modal.Header>Enter Your Details</Modal.Header>
+        <Modal.Content>
+          {/* <RegistrationForm /> */}
```
Comment (FlisAnn) [inline src/components/MasterModal.jsx]: This is commented out?

**Review by FlisAnn — COMMENTED** [review]
Just some commented out code left, otherwise it looks good!

File: src/components/MasterModal.jsx (hunk 3 above)
Comment (sealfury) [reply src/components/MasterModal.jsx]: Removed the commented out code in both instances

File: cypress/integration/visitorCanRegisterForAnAccount.feature.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  });
+
+  describe("not register with invalid credentials", () => {
+    beforeEach(() => {
+      cy.route({
+        method: "POST",
+        url: "http://localhost:3000/api/**",
+        status: "401",
+        response: {
+          errors: "Registration unsuccessful",
```
Comment (emtalen) [inline cypress/integration/visitorCanRegisterForAnAccount.feature.js]: Look at devise token auth's documentation and update this error message and the way you receive it

File: src/components/DisplayArticlesData.jsx [hunk 5]
```diff
@@ -21,7 +21,8 @@ const DisplayArticlesData = () => {
   );
 
   return (
-    <>{mainArticles.length > 0 && <div data-cy="index">{articleIndex}</div>}</>
+
```
Comment (emtalen) [inline src/components/DisplayArticlesData.jsx]: remove this extra space here

File: src/components/MasterModal.jsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+          size="medium"
+        >
+          <Modal.Header data-cy="header-user-email">
+            <h3>Logged in as {currentUser.email}</h3>
+          </Modal.Header>
+          <Modal.Header>Enter Card Details</Modal.Header>
+          <Modal.Content>Card payment form goes here</Modal.Content>
+          <Modal.Description></Modal.Description>
+          <Modal.Actions>
+            <Button icon="check" content="All Done" onClick={finalizePayment} />
```
Comment (emtalen) [inline src/components/MasterModal.jsx:74]: I would argue this is a liiiittle bit outside of the scope for this feature

File: src/components/MenuHeader.jsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        fluid="true" />
+        <Menu.Item 
+        />
+        <Menu.Item position="right" >
+          <MasterModal />
+        </Menu.Item>
+      </Menu>
+    </Segment>
+  )
+
```
Comment (emtalen) [inline src/components/MenuHeader.jsx]: remove extra line

File: src/index.js [hunk 8]
```diff
@@ -8,17 +8,20 @@ import { Provider } from 'react-redux';
 import 'semantic-ui-css/semantic.min.css';
 import { BrowserRouter } from "react-router-dom"
 
-axios.defaults.baseURL = process.env.REACT_APP_API_URL
- 
+axios.defaults.baseURL =
+  process.env.NODE_ENV === "production"
+    ? "https://dream-time-news-api.herokuapp.com"
+    : "http://localhost:3000/api";
```
Comment (emtalen) [inline src/index.js]: please go back to use an .env file

File: cypress/fixtures/visitor_can_register.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
+  "success": true,
```
Comment (emtalen) [inline cypress/fixtures/visitor_can_register.json]: please double check that this is really part of the response and received as a boolean

**Review by emtalen — DISMISSED** [review]

File: cypress/fixtures/visitor_can_register.json (hunk 9 above)
Comment (sealfury) [reply cypress/fixtures/visitor_can_register.json]: Have added logic to convert the response to boolean and have updated the fixture to reflect the actual response

File: src/components/MasterModal.jsx (hunk 6 above)
Comment (sealfury) [reply src/components/MasterModal.jsx:74]: Agreed, could have just as well called that `{closeBoth}` for now but I figured it served as a good indicator of how to proceed with implementation on the payment feature

File: cypress/integration/visitorCanRegisterForAnAccount.feature.js (hunk 4 above)
Comment (sealfury) [reply cypress/integration/visitorCanRegisterForAnAccount.feature.js]: Updated

File: src/components/MenuHeader.jsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    <Segment inverted>
+      <Menu inverted>
+        <Menu.Item 
+          as={Link}
+          to={{pathname: "/"}}
+        />
+        <Image 
+        src="./assets/dreamtimesnews.png"
+        size="large"
+        fluid="true" />
```
Comment (tochman) [inline src/components/MenuHeader.jsx]: Please fix code formatting

File: src/components/SingleArticle.jsx [hunk 11]
```diff
@@ -21,19 +21,18 @@ const SingleArticle = () => {
...
           <p data-cy="article-author">{singleArticle.author}</p>
         </Grid.Column>
       </Grid>
-      <Button data-cy="back-button" as={NavLink} to="/">
+      <Button data-cy="back-button" as={NavLink} to="/" color="twitter">
         Back
       </Button>
       {errorMessage && (
-        <Container data-cy="error-message-article">
+        <Message color="red" data-cy="error-message-article">
```
Comment (tochman) [inline src/components/SingleArticle.jsx]: I would suggest 

```
  <Message
    header={errorMessage}
    data-cy="error-message-article"
  />
  ```
No need to pass in any html

**Review by tochman — APPROVED** [review]
Okay, finally...

**Pull request merged**